realtek: add support for kernel 6.18

jonasjelonek opened the pull request:
Making this public quite early so everyone knows that this is being worked on, and that we can catch major issues earlier.

---

This adds support for kernel 6.18 to realtek target, refreshes/drops/fixes patches and adjusts our downstream drivers.

First tests show no major issues (in contrast to what we had with 6.6 and 6.12).

Tested:
- rtl838x with Zyxel GS1900-10HP B1
- rtl930x with Zyxel XGS1210-12 A1, XikeStor SKS8310-8X, TP-Link TL-ST1008F

Known issues:
- Aquantia PHY not working right now

Any testing/input is appreciated :)

jonasjelonek committed:
realtek: add v6.18 as testing kernel

Add kernel v6.18 as testing kernel for realtek target.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21181
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek committed:
realtek: add missing symbols for 6.18

Add missing symbols to default config by running make kernel_oldconfig
and using default options for all unanswered symbols.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21181
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek committed:
realtek: phy: 6.18: drop shared priv access helper

Between v6.12 and v6.18, struct phy_package_shared was moved from
include/linux/phylib.h to drivers/net/phy/phy_package.c and thus made
internal/unavailable for other drivers.

Our PHY driver used a helper to access the shared priv. With v6.18, it
doesn't work anymore (due to missing access to internal fields of struct
phy_package_shared) but upstream has an equivalent helper for that. Drop
ours and use the upstream helper.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21181
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek committed:
realtek: pcs: 6.18: fix neg_mode handling

Between 6.12 and 6.18, neg_mode moved from struct phylink_pcs to a
parameter passed to pcs_get_state. Adjust the implementation
accordingly.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21181
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek committed:
realtek: dsa: 6.18: adjust mac ops and callbacks

Between 6.12 and 6.18, several DSA switch ops have changed.
phylink_mac_* ops were moved out of dsa_switch_ops into dsa_switch using
phylink_mac_ops. Add new definitions for that and adjust the callbacks
accordingly to work with the different signatures.

get_mac_eee() has been dropped and support_eee() was added.

remove_new() has been dropped completely, allowing us to use remove()
again.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21181
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul reviewed (changes requested):
Unbelievable how less work is left in contrast to 6.6 and 6.12.

jonasjelonek commented:
> Unbelievable how less work is left in contrast to 6.6 and 6.12.

Those credits go to you :)

> Could you cherry-pick Haukes revert-patch for the time being? Verified this with 6.12 already. We could probably drop this later one again.

Sure, I'll include it in the next push

jonasjelonek committed:
realtek: phy: 6.18: switch to sfp module caps

Kernel 6.18 uses dedicated sfp_module_caps instead of a plain bitmask
for supported modes. Switch to that the fix the rtl83xx-phy driver.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21181
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul commented:
@jonasjelonek As long as 6.18 is not merged in https://github.com/openwrt/openwrt/pull/20965 can you please add those patches to this PR. This avoids inconsistencies while both patchsets evolve ...

plappermaul commented:
Fixed. Added another source/origin to get the PR directly. 

Nevertheless: For the time being please remove pending-6.18/487-mtd-spinand-Add-support-for-Etron-EM73D044VCx.patch This fixes our nand builds. This issue is already mentioned in the generic/malta PR.

plappermaul commented on target/linux/realtek/files-6.18/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -59,7 +60,9 @@ static const struct firmware rtl838x_8218b_fw;
 static inline struct phy_device *get_package_phy(struct phy_device *phydev, int port)
 {
-	return mdiobus_get_phy(phydev->mdio.bus, phydev->shared->base_addr + port);
+	int base_addr = phydev->mdio.addr & ~7;
```
Don't like that because we have 4 port PHYs like the RTL8214FC. Please check https://github.com/openwrt/openwrt/pull/21204

jameywine commented:
Hi @jonasjelonek 

I have tested on my, not yet supported SoC, rtl9607c based router and it boots just fine (just had to move all the things from [my pull request](https://github.com/openwrt/openwrt/pull/20064) to 6.18, thats it). I don't know if that helps here but testing is still testing. At least i know that i can easily rebase my pr to add 6.18 kernel support if this pr gets merged first.
<details> 
<summary>Linux 6.18 boot log</summary>

```ini
## Booting kernel from Legacy Image at 83000000 ...
   Image Name:   MIPS OpenWrt Linux-6.18.2
   Image Type:   MIPS Linux Kernel Image (lzma compressed)
   Data Size:    5409595 Bytes = 5.2 MiB
   Load Address: 80100000
   Entry Point:  80100000
   Verifying Checksum ... OK
   Uncompressing Kernel Image
[    0.000000] Linux version 6.18.2 (binguy@binpc) (mips-openwrt-linux-musl-gcc (OpenWrt GCC 14.3.0 r32359+35-22a69dfa4a) 14.3.0, GNU ld (GNU Binutils) 2.44) #0 SMP Sun Dec 21 12:43:26 2025
[    0.000000] SoC Type: Realtek RTL9607C subtype 18 rev B (6831)
[    0.000000] printk: legacy bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 0001a120 (MIPS interAptiv (multi))
[    0.000000] MIPS: machine is BT-PON BT-G711AX
[    0.000000] earlycon: ns16550a0 at MMIO 0x18002000 (options '115200n8')
[    0.000000] printk: legacy bootconsole [ns16550a0] enabled
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Using appended Device Tree.
[    0.000000] OF: reserved mem: Reserved memory: No reserved-memory node in the DT
[    0.000000] VPE topology {2,2} total 4
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.000000] MIPS secondary cache 256kB, 8-way, linesize 32 bytes.
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x000000000fffffff]
[    0.000000]   HighMem  empty
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x000000000fffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x000000000fffffff]
[    0.000000] percpu: Embedded 12 pages/cpu s18640 r8192 d22320 u49152
[    0.000000] pcpu-alloc: s18640 r8192 d22320 u49152 alloc=12*4096
[    0.000000] pcpu-alloc: [0] 0 [0] 1 [0] 2 [0] 3 
[    0.000000] Kernel command line: earlycon
[    0.000000] printk: log buffer data + meta data: 131072 + 409600 = 540672 bytes
[    0.000000] Dentry cache hash table entries: 32768 (order: 5, 131072 bytes, linear)
[    0.000000] Inode-cache hash table entries: 16384 (order: 4, 65536 bytes, linear)
[    0.000000] Writing ErrCtl register=00000000
[    0.000000] Readback ErrCtl register=00000000
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 65536
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=4, Nodes=1
[    0.000000] rcu: Hierarchical RCU implementation.
[    0.000000]  Tracing variant of Tasks RCU enabled.
[    0.000000] rcu: RCU calculated value of scheduler-enlistment delay is 10 jiffies.
[    0.000000] RCU Tasks Trace: Setting shift to 2 and lim to 1 rcu_task_cb_adjust=1 rcu_task_cpu_ids=4.
[    0.000000] NR_IRQS: 256
[    0.000000] rcu: srcu_init: Setting srcu_struct sizes based on contention.
[    0.000000] Failed to get CPU clock: -2
[    0.000000] CPU frequency from device tree: 1150MHz
[    0.000000] clocksource: GIC: mask: 0xffffffffffffffff max_cycles: 0x10939a911d9, max_idle_ns: 440795210078 ns
[    0.000003] sched_clock: 64 bits at 1150MHz, resolution 0ns, wraps every 2199023255551ns
[    0.009262] clocksource: realtek_otto_timer: mask: 0xfffffff max_cycles: 0xfffffff, max_idle_ns: 38225208801 ns
[    0.020677] Calibrating delay loop... 761.03 BogoMIPS (lpj=3805184)
[    0.067552] pid_max: default: 32768 minimum: 301
[    0.080877] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.089046] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.112018] rcu: Hierarchical SRCU implementation.
[    0.117387] rcu:     Max phase no-delay instances is 1000.
[    0.124778] smp: Bringing up secondary CPUs ...
[    0.131119] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.131159] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.131170] MIPS secondary cache 256kB, 8-way, linesize 32 bytes.
[    0.131249] CPU1 revision is: 0001a120 (MIPS interAptiv (multi))
[    0.171370] Counter synchronization [CPU#0 -> CPU#1]: passed
[    0.209488] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.209527] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.209538] MIPS secondary cache 256kB, 8-way, linesize 32 bytes.
[    0.209585] CPU2 revision is: 0001a120 (MIPS interAptiv (multi))
[    0.258492] Counter synchronization [CPU#0 -> CPU#2]: passed
[    0.295732] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.295767] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.295778] MIPS secondary cache 256kB, 8-way, linesize 32 bytes.
[    0.295833] CPU3 revision is: 0001a120 (MIPS interAptiv (multi))
[    0.335961] Counter synchronization [CPU#0 -> CPU#3]: passed
[    0.371882] smp: Brought up 1 node, 4 CPUs
[    0.378299] Memory: 235688K/262144K available (8379K kernel code, 647K rwdata, 1720K rodata, 11324K init, 244K bss, 25080K reserved, 0K cma-reserved, 0K highmem)
[    0.397523] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.408828] posixtimers hash table entries: 2048 (order: 2, 16384 bytes, linear)
[    0.417304] futex hash table entries: 1024 (32768 bytes on 1 NUMA nodes, total 32 KiB, linear).
[    0.432019] pinctrl core: initialized pinctrl subsystem
[    0.440419] NET: Registered PF_NETLINK/PF_ROUTE protocol family
[    0.447655] thermal_sys: Registered thermal governor 'step_wise'
[    0.447963] FPU Affinity set after 15220 emulations
[    0.472771] clocksource: Switched to clocksource realtek_otto_timer
[    0.487956] NET: Registered PF_INET protocol family
[    0.493714] IP idents hash table entries: 4096 (order: 3, 32768 bytes, linear)
[    0.502845] tcp_listen_portaddr_hash hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.512211] Table-perturb hash table entries: 65536 (order: 6, 262144 bytes, linear)
[    0.520887] TCP established hash table entries: 2048 (order: 1, 8192 bytes, linear)
[    0.529495] TCP bind hash table entries: 2048 (order: 3, 32768 bytes, linear)
[    0.537625] TCP: Hash tables configured (established 2048 bind 2048)
[    0.545702] MPTCP token hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.554181] UDP hash table entries: 256 (order: 2, 14336 bytes, linear)
[    0.561645] UDP-Lite hash table entries: 256 (order: 2, 14336 bytes, linear)
[    0.570525] NET: Registered PF_UNIX/PF_LOCAL protocol family
[    0.578918] workingset: timestamp_bits=14 max_order=16 bucket_order=2
[    0.589200] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.595925] jffs2: version 2.2 (NAND) (SUMMARY) (ZLIB) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.628411] Serial: 8250/16550 driver, 2 ports, IRQ sharing disabled
[    0.637437] printk: legacy console [ttyS0] disabled
[    0.644512] 18002000.uart: ttyS0 at MMIO 0x18002000 (irq = 22, base_baud = 12500000) is a 16550A
[    0.654577] printk: legacy console [ttyS0] enabled
[    0.654577] printk: legacy console [ttyS0] enabled
[    0.665337] printk: legacy bootconsole [early0] disabled
[    0.665337] printk: legacy bootconsole [early0] disabled
[    0.677542] printk: legacy bootconsole [ns16550a0] disabled
[    0.677542] printk: legacy bootconsole [ns16550a0] disabled
[    0.723105] brd: module loaded
[    0.730790] spi-nand spi0.0: Winbond SPI NAND was found.
[    0.737101] spi-nand spi0.0: 128 MiB, block size: 128 KiB, page size: 2048, OOB size: 64
[    0.747020] 5 fixed-partitions partitions found on MTD device spi0.0
[    0.754368] Creating 5 MTD partitions on "spi0.0":
[    0.759784] 0x000000000000-0x0000000e0000 : "u-boot"
[    0.769420] 0x0000000e0000-0x000000100000 : "u-boot-env"
[    0.777601] 0x000000100000-0x000000120000 : "u-boot-env2"
[    0.787445] 0x000000120000-0x000000140000 : "static_conf"
[    0.798392] 0x000000140000-0x000007d80000 : "ubi"
[    0.911730] i2c_dev: i2c /dev entries driver
[    0.926762] NET: Registered PF_INET6 protocol family
[    0.935661] Segment Routing with IPv6
[    0.939840] In-situ OAM (IOAM) with IPv6
[    0.944459] NET: Registered PF_PACKET protocol family
[    0.950179] 8021q: 802.1Q VLAN Support v1.8
[    0.989912] UBI: auto-attach mtd4
[    0.993765] ubi0: attaching mtd4
[    1.702836] random: crng init done
[    1.933224] ubi0: scanning is finished
[    2.034770] ubi0: attached mtd4 (name "ubi", size 124 MiB)
[    2.041033] ubi0: PEB size: 131072 bytes (128 KiB), LEB size: 126976 bytes
[    2.048793] ubi0: min./max. I/O unit sizes: 2048/2048, sub-page size 2048
[    2.056410] ubi0: VID header offset: 2048 (aligned 2048), data offset: 4096
[    2.064194] ubi0: good PEBs: 994, bad PEBs: 0, corrupted PEBs: 0
[    2.070888] ubi0: user volume: 5, internal volumes: 1, max. volumes count: 128
[    2.078951] ubi0: max/mean erase counter: 5/2, WL threshold: 4096, image sequence number: 163500012
[    2.089052] ubi0: available PEBs: 303, total reserved PEBs: 691, PEBs reserved for bad PEB handling: 20
[    2.099598] ubi0: background thread "ubi_bgt0d" started, PID 355
[    2.099750] clk: Disabling unused clocks
[    2.163099] Freeing unused kernel image (initmem) memory: 11324K
[    2.169830] This architecture does not have kernel memory protection.
[    2.177046] Run /init as init process
[    2.181125]   with arguments:
[    2.184436]     /init
[    2.186960]   with environment:
[    2.190444]     HOME=/
[    2.193078]     TERM=linux
[    2.588355] init: Console is alive
[    2.592561] init: - watchdog -
[    2.607933] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    2.616595] gpio_button_hotplug: loading out-of-tree module taints kernel.
[    2.626094] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    2.638663] init: - preinit -
ls: /sys/class/net/lan*: No such file or directory
BusyBox v1.37.0 (2025-12-21 12:43:26 UTC) multi-call binary.

Usage: basename FILE [SUFFIX] | -a FILE... | -s SUFFIX FILE...

Strip directory path and SUFFIX from FILE

        -a              All arguments are FILEs
        -s SUFFIX       Remove SUFFIX (implies -a)
[    2.981381] mtdblock: MTD device 'u-boot-env2' is NAND, please consider using UBI block devices instead.
[    3.146226] mtdblock: MTD device 'u-boot-env2' is NAND, please consider using UBI block devices instead.
[    3.311961] mtdblock: MTD device 'u-boot-env' is NAND, please consider using UBI block devices instead.
Cannot parse config file '/etc/fw_env.config': No such file or directory
Failed to find NVMEM device
Press the [f] key and hit [enter] to enter failsafe mode
Press the [1], [2], [3] or [4] key and hit [enter] to select the debug level
[    7.823065] procd: - early -
[    7.826630] procd: - watchdog -
[    8.410796] procd: - watchdog -
[    8.414758] procd: - ubus -
[    8.473005] procd: - init -
Please press Enter to activate this console.
[    8.891783] kmodloader: loading kernel modules from /etc/modules.d/*
[    8.963232] kmodloader: done loading kernel modules from /etc/modules.d/*
[    9.684542] urngd: v1.0.2 started.
[   10.369823] mtdblock: MTD device 'u-boot-env2' is NAND, please consider using UBI block devices instead.
[   10.533729] mtdblock: MTD device 'u-boot-env2' is NAND, please consider using UBI block devices instead.
[   10.699603] mtdblock: MTD device 'u-boot-env' is NAND, please consider using UBI block devices instead.



BusyBox v1.37.0 (2025-12-21 12:43:26 UTC) built-in shell (ash)

  _______                     ________        __
 |       |.-----.-----.-----.|  |  |  |.----.|  |_
 |   -   ||  _  |  -__|     ||  |  |  ||   _||   _|
 |_______||   __|_____|__|__||________||__|  |____|
          |__| W I R E L E S S   F R E E D O M
 -----------------------------------------------------
 OpenWrt SNAPSHOT, r32359+35-22a69dfa4a
 -----------------------------------------------------

 === WARNING! =====================================
 There is no root password defined on this device!
 Use the "passwd" command to set up a new password
 in order to prevent unauthorized SSH logins.
 --------------------------------------------------


 OpenWrt recently switched to the "apk" package manager!

 OPKG Command           APK Equivalent      Description
 ------------------------------------------------------------------
 opkg install <pkg>     apk add <pkg>       Install a package
 opkg remove <pkg>      apk del <pkg>       Remove a package
 opkg upgrade           apk upgrade         Upgrade all packages
 opkg files <pkg>       apk info -L <pkg>   List package contents
 opkg list-installed    apk info            List installed packages
 opkg update            apk update          Update package lists
 opkg search <pkg>      apk search <pkg>    Search for packages
 ------------------------------------------------------------------

For more information visit:
https://openwrt.org/docs/guide-user/additional-software/opkg-to-apk-cheatsheet

root@OpenWrt:~#
```

</details>

namiltd commented:
Please sync your PR with the current one https://github.com/openwrt/openwrt/pull/20965.

jonasjelonek committed:
realtek: add patch to extend upstream sfp smbus support

Add patches that extends the upstream smbus support in sfp driver. This
only supports SMBus byte operations and in that case, disables hwmon and
prints warnings because of that. Using the word/block capability of our
I2C controller is totally fine for all that functionality.

The patches are currently pending upstream.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21181
Signed-off-by: Robert Marko <robimarko@gmail.com>

Neustradamus commented:
@jonasjelonek: Thanks for your work!

jonasjelonek commented:
encountered kind of a weird issue that I want to note here ... running one Realtek switch with 6.12 on the one end, and another one with 6.18 at the other end, doesn't seem to play that nicely together.

Running 6.18 on a TP-Link ST1008F and 6.12 on a XikeStor SKS8310-8X, a 1000Base-X link doesn't come up properly. It is shown as `UP` on the 6.18 switch, but as `LOWERLAYERDOWN` on the 6.12 switch. Traffic doesn't flow. As soon as I run 6.18 on the other switch too, it works.

No obvious idea why, but definitely something to test with 6.18 as testing kernel, and something to keep in mind until everything is on 6.18.

GoetzGoerisch commented:
for the record:
running 6.18 since commit 9e3adf0 with an uptime of now 19d on my desktop switch with 3 devices and a couple of VLANs
Thank you all! 👍🏻

jonasjelonek committed:
realtek: 6.18: remove obsolete patches

Remove patches which are accepted upstream and were backported to
pre-6.18. Those are obsolete now.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21181
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek committed:
realtek: 6.18: adapt SFP SMBus patches

Adapt our downstream patches for SMBus SFP functionality to kernel 6.18.
Upstream gained support for SMBus byte access on SFP and a patch to
extend that to word and block access is pending.

These downstream patches aren't acceptable for upstream and will need to
be adapted. However, keep them for now to maintain functionality.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21181
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek committed:
realtek: 6.18: adapt and refresh remaining patches

Refresh and adapt the remaining patches as needed.

- 024-02-v7.1-net-phy-realtek-add-RTL8224-pair-order-support.patch
  needed addition of '#include "../phylib.h"'
- 314-irqchip-irq-realtek-rtl-add-VPE-support.patch needed manual
adoption
- 700-dsa-mdio-increase-max-ports-for-rtl839x-rtl931x.patch needed
  manual fix
- 810-lm75-alert-polarity-swap.patch needed adoption to changed probe
  function upstream

All other patches were automatically refreshed.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21181
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek committed:
kernel/realtek: create files for v6.18 (from v6.12)

This is an automatically generated commit.

When doing `git bisect`, consider `git bisect --skip`.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21181
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek committed:
kernel/realtek: restore files for v6.12

This is an automatically generated commit which aids following Kernel patch
history, as git will see the move and copy as a rename thus defeating the
purpose.

For the original discussion see:
https://lists.openwrt.org/pipermail/openwrt-devel/2023-October/041673.html

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21181
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek commented:
Having a few issues right now (e.g. 10G-T ports on XGS1250-12 A1 not working correctly), I'm investigating them. Shouldn't be a blocker for adding 'testing' support though.

jonasjelonek commented:
Current guess is that this seems specific to Aquantia PHYs because:
- my XGS1250-12 A1 has those
- trying on an RTL931x device, I first forgot to include the Aquantia PHY driver to the kernel config. PHYs were detected as "Generic C45 PHY" and it worked. Then I added Aquantia driver and now they are not working anymore

Can someone with e.g. a XGS1250-12 B1 with the Realtek PHYs on the 10G copper ports confirm my guess (they should still work fine)?

plappermaul commented:
successfully loop-back Tested on XGS1210 (RTL930x): 
- 1G on RTL8218D only
- 1G on RTL8218D/RTL8226
- 2.5G on RTL8226
- 1G/10G on SFP

small observation during first boot (never seen afterwards)

```
[    2.768859] RTL8226-CG 2.5Gbps PHY realtek-mdio:18: rtl822x_c45_soft_reset failed: -145
[    2.779546] rtl83xx-switch 1b000000.switchcore:ethernet-switch lan9 (uninitialized): failed to connect to PHY: -ETIMEDOUT
[    2.791878] rtl83xx-switch 1b000000.switchcore:ethernet-switch lan9 (uninitialized): error -145 setting up PHY for tree 0, switch 0, port 24
```

plappermaul commented:
successfully loop-back tested on Linksys DGS-1210-28 (RTL838x):

- 1G on RTL8218B only
- 1G on RTL8218B/RTL8214FC
- 1G on RTL8214FC only
- 1G on SFP

minor issue detected https://github.com/openwrt/openwrt/issues/22658
